### ISSUE TITLE
Feature: Version a log to allow compacting

### ISSUE THREAD

**Issue by philholden** [substack/swarmlog#5]
I am using a log to maintain a document store. Documents can be added edited and removed and these actions are stored into log. Each client then reduces the log to reconstruct the store. However some of these actions are quite verbose eg editing just adds the document again with different text but the same id.

It would be nice if I could compact a log so it only has the latest version of each document and then increment a version number. Other clients would know not to sync out of date versions and scrub previous versions if a new version is available.

I know I can create new log but I would be keen to keep the same private and public keys as the data being presented does not change.

**Comment by elsehow** [comment]
you might be interested in https://github.com/maxogden/dat

if you really need to, you can roll this yourself. hyperlog and swarmlog
have a `db` property, which has a method `del(key, cb)`

**Comment by philholden** [comment]
@elsehow thanks what will happen when this syncs with other clients will they delete this node from the log too?

Sorry this is unrelated but how can I generate `ssbKeys` on the client? I want each client to generate them and put them in localstorage if not there already. Get lots of errors when I import `ssb-keys` via webpack. I don't want a server.

**Comment by elsehow** [comment]
@philholden no, in fact, these clients will replicate those nodes to you! but, if all clients are on the same page about which ones to delete, then it shouldn't be (too much of) a problem.

note that "being on the same page about which ones to delete" is non-trivial. sequence numbers are not globally unique, and timestamps bring along their own problems with clock synchronization. a mechanism that relies on the causal linking structure of hyperlogs is most reliable - you might want to look into vector clocks.

if you are just looking for a distribute key-value store that any client can update, [hyperkv](https://github.com/substack/hyperkv) may be for you.

as far as generating `ssbKeys` - i believe these keys are just ed25519 keys, which you should be able to generate with chloride/browser. though, you should verify this. if it works, let us know, and/or post a module that generates ssb keys in the browser!

**Comment by philholden** [comment]
@elsehow I am actually using `redux` for clientside state and using `swarmlog` to sync `actions` across clients. I am persisting the database using an `IndexDB`.

Looks like `chloride/browser` does not have methods for generating elliptic curves keys. But I found `supercop.js` which works:

```javascript
import supercop from 'supercop.js'

export default function generateKeys() {
  const keys = supercop.createKeypair()
  return {
    public: `${keys.pubKey.toString('base64')}.ed25519`,
    private: `${keys.privKey.toString('base64')}.ed25519`
  }
}
```

How does Swarmlog scale i.e. if I have 100,000 users subscribing to the same log will it do something intelligent so every client is only connected directly to a handful of other clients.

Thanks for your help. Where is a good place to get help `hyperlog`, `swarmlog` etc is there a channel somewhere?

**Comment by elsehow** [comment]
> How does Swarmlog scale i.e. if I have 100,000 users subscribing to the same log will it do something intelligent so every client is only connected directly to a handful of other clients.

i believe signalhubs are smart, and make sure that there is *some* path between any two given users in a swarm, without connecting everyone to everyone (though you should verify this with @substack or @mafintosh). 

another nice thing is, swarmlog will take many signalhubs, so we don't have to rely on any single running service.

i have asked questions on both #scuttlebutt and #dat on freenode. those projects focus around other things, but folks do hang out there.

**Comment by elsehow** [comment]
also, using hyperlogs to persist as sync application state is cool to me. i have long been thinking about modelling application state as an append-only log of atomic events, where the state is some reduce over a global log of actions.

do note the question of ordering though. if you do `log.createReadStream()`, you may see a different order of events that hit the redux store on different clients.

**Comment by philholden** [comment]
Thanks this is awesome you can launch something and it goes global over night and don't have to think about load balancers, hosting, cloud & containers.
